ansible/roles/docker/tasks/main.yml

```
@@ -41,26 +42,63 @@
 
   # tcp socket service requires docker service to be started after it
-- name: start docker tcp socket service
-  shell: sudo systemctl stop docker && sudo systemctl start docker-tcp.socket && sudo systemctl start docker
-  when: (docker_tcp_socket | changed) or not docker_tcp_socket_state.stdout | match("Active.* active")
+- name: reload systemd configuration
+  shell: sudo systemctl daemon-reload
+  when: "(docker_tcp_socket | changed) or (docker_tcp_socket_state.stdout != 'Active: active')"
+
+- name: stop docker
+  service:
+    name: docker
+    state: stopped
+  when: "(docker_tcp_socket | changed) or (docker_tcp_socket_state.stdout != 'Active: active')"
+
+- name: start docker-tcp service
+  service:
+    name: docker-tcp.socket
+    state: started
+  when: "(docker_tcp_socket | changed) or (docker_tcp_socket_state.stdout != 'Active: active')"
+
+- name: check docker service state
+  shell: sudo systemctl status docker | grep 'Active.*active' -o
+  ignore_errors: true
+  register: docker_service_state
+  tags:
+    - prebake-for-dev
 
 - name: remove the docker key file, if any. It shall be regenerated by docker on restart
   file: name=/etc/docker/key.json  state=absent
 
 # XXX: service module doesn't do daemon-reload yet, so need to use shell module here
 # https://github.com/ansible/ansible-modules-core/issues/191
-- name: restart docker
+- name: reload docker systemd configuration
   #service: name=docker state=restarted
-  shell: sudo systemctl daemon-reload && sudo systemctl restart docker
+  shell: sudo systemctl daemon-reload
+  when: "(docker_service_state.stderr | match('.*docker.service changed on disk.*')) or (docker_service_state.stdout != 'Active: active')"
   tags:
     - prebake-for-dev
 
+# XXX: this needs to happen twice after setting up the thin provisioner because
+#      of some docker bug I've not investigated.
+- name: restart docker (first time)
+  service:
+    name: docker
+    state: restarted
+  when: thin_provisioned|changed
```

hmm, I think the error message I pasted is causing the task to be not run (it fails and failure is ignored) so restart is actually not triggered. We should fix that I think.